File: frameworks/python/django-asgi/Implementation/views/middleware.py
from django.conf import settings
from django.http import JsonResponse
from django.utils.decorators import decorator_from_middleware
from django.utils.deprecation import MiddlewareMixin
from django.views.decorators.http import require_GET

P = settings.PAYLOADS


# rb:wiring middleware.*
class Noop(MiddlewareMixin):
    """One layer. Returning None from process_request carries the request on to the next layer."""

    def process_request(self, request):
        return None


def layers(count: int):
    """Django's MIDDLEWARE runs on every request. decorator_from_middleware is how Django scopes a
    middleware to one view, as cache_page and gzip_page scope theirs, so each layer is Noop made
    into a view decorator."""
    layer = decorator_from_middleware(Noop)

    def wrap(view):
        for _ in range(count):
            view = layer(view)
        return view

    return wrap
# rb:end


# middleware: no-op layers in front of the view.
# rb:handler middleware.none
@require_GET
async def none(request):
    return JsonResponse(P.small)


# rb:handler middleware.four
@require_GET
@layers(4)
async def four(request):
    return JsonResponse(P.small)


# rb:handler middleware.sixteen
@require_GET
@layers(16)
async def sixteen(request):
    return JsonResponse(P.small)

File: frameworks/go/huma/Implementation/middleware.go
package implementation

import (
	"context"

	"github.com/danielgtaylor/huma/v2"
)

// middlewareRoutes put no-op layers in front of the handler, as Huma middleware on the operation
// itself, which runs after router middleware and before the handler.
func middlewareRoutes(api huma.API, p *Payloads) {
	small := func(ctx context.Context, _ *struct{}) (*PayloadOutput, error) {
		return &PayloadOutput{Body: &p.Small}, nil
	}

	huma.Get(api, "/middleware/none", small)

	huma.Get(api, "/middleware/four", small, func(o *huma.Operation) { o.Middlewares = layered(4) })

	huma.Get(api, "/middleware/sixteen", small, func(o *huma.Operation) { o.Middlewares = layered(16) })
}

// rb:wiring middleware.*
// noop is one layer: it calls the next and does nothing else.
func noop(ctx huma.Context, next func(huma.Context)) {
	next(ctx)
}

// layered is count no-op layers, for an operation's Middlewares.
func layered(count int) huma.Middlewares {
	layers := make(huma.Middlewares, count)
	for i := range layers {
		layers[i] = noop
	}
	return layers
}

// rb:end
